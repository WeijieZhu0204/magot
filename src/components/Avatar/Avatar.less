@import '../../variable.less';

.@{prefix}-avatar {
  .box-sizing();
  .box();

  position: relative;
  display: inline-block;
  background-color: @avatar-background-color;
  color: @avatar-color;
  font-weight: 400;
  overflow: hidden;
  white-space: nowrap;
  text-align: center;
  vertical-align: middle;

  .avatar-size(@size-base; @avatar-icon-font-size-base;);
}

.@{prefix}-avatar-sm {
  .avatar-size(@size-sm; @avatar-icon-font-size-sm;);
}

.@{prefix}-avatar-lg {
  .avatar-size(@size-lg; @avatar-icon-font-size-lg;);
}

.@{prefix}-avatar-image {
  background: none;
  & > img {
    display: block;
    width: 100%;
    height: 100%;
  }
}

.@{prefix}-avatar-text {
  & > .text {
    position: absolute;
    left: 50%;
    transform-origin: 0 center;
  }
}

.avatar-size(@size; @font-size) {
  width: @size;
  height: @size;
  line-height: @size;
  &.@{prefix}-avatar-icon {
    font-size: @font-size;
  }
}
