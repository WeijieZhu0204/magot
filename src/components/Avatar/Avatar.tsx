import React from 'react';
import Icon from '../Icon';
import * as component from '../component';
import { getShallowTextSize } from '../../utils/node';

import './Avatar.less';

export interface AvatarProps
  extends component.BaseComponent,
    component.SizedComponent {
  /**
   * 图片头像的图片地址
   */
  src?: string;

  /**
   * 多图适配的图片地址，仅适用于图片头像
   */
  srcset?: string;

  /**
   * 当图片加载失败时的占位文案，仅适用于图片头像
   */
  alt?: string;

  /**
   * 文字头像的文字内容
   */
  text?: string;

  /**
   * 图标头像的图标名
   * @default user
   */
  icon?: string;

  /**
   * 头像的圆角，默认全圆角
   * @default 50%
   */
  radius?: number | string;
}

const defaultProps: Partial<AvatarProps> = {
  ...component.getDefaultSizedProps(),
  icon: 'user',
  radius: '50%',
};

function Avatar(props: AvatarProps) {
  let children;
  let isImageAvatar = false;
  let isTextAvatar = false;
  let isIconAvatar = false;

  if (props.src) {
    isImageAvatar = true;
    children = <AvatarImage {...props} />;
  } else if (props.text) {
    isTextAvatar = true;
    children = <AvatarText {...props} text={props.text} />;
  } else if (props.icon) {
    isIconAvatar = true;
    children = <Icon className="icon" name={props.icon} />;
  } else {
    children = null;
  }

  const type = 'avatar';
  const prefix = component.getComponentPrefix(type);
  const cls = component.getComponentClasses(type, props, {
    [prefix + '-image']: isImageAvatar,
    [prefix + '-text']: isTextAvatar,
    [prefix + '-icon']: isIconAvatar,
  });

  return (
    <div
      className={cls}
      style={{
        borderRadius: props.radius,
        ...props.style,
      }}>
      {children}
    </div>
  );
}

function AvatarImage(props: AvatarProps) {
  return (
    <img
      className="image"
      src={props.src}
      alt={props.alt}
      srcSet={props.srcset}
    />
  );
}

function AvatarText(props: AvatarProps & { text: string }) {
  const avatarSizes = { small: 24, normal: 32, large: 40 };
  const avatarWidth = avatarSizes[props.size || 'normal'];
  const { width: textWidth } = getShallowTextSize(props.text);
  const textScale = Math.min(1, (avatarWidth - 6) / textWidth);
  return (
    <span
      className="text"
      style={{ transform: `scale(${textScale}) translateX(-50%)` }}>
      {props.text}
    </span>
  );
}

Avatar.defaultProps = defaultProps;

export default Avatar;
