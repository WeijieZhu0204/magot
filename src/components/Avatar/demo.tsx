import React from 'react';
import Avatar from './Avatar';
import { Link } from 'react-router-dom';

function AvatarDemo() {
  return (
    <>
      <Link to="/" className="demo-goback">
        返回
      </Link>
      <div className="demo-box">
        <Avatar />
        <Avatar icon="user" />
        <Avatar text="USER" />
        <Avatar src={require('./avatar.png')} />
        <Avatar icon="user" style={{ backgroundColor: '#52c41a' }} />
      </div>
      <div className="demo-box">
        <Avatar radius={4} />
        <Avatar icon="user" radius={4} />
        <Avatar text="USER" radius={4} />
        <Avatar src={require('./avatar.png')} radius={4} />
        <Avatar icon="user" radius={4} style={{ backgroundColor: '#52c41a' }} />
      </div>
      <div className="demo-box">
        <Avatar icon="user" size="small" />`
        <Avatar icon="user" size="normal" />
        <Avatar icon="user" size="large" />
        <Avatar text="USER" size="small" />
        <Avatar text="USER" size="normal" />
        <Avatar text="USER" size="large" />
        <Avatar src={require('./avatar.png')} size="small" />
        <Avatar src={require('./avatar.png')} size="normal" />
        <Avatar src={require('./avatar.png')} size="large" />
      </div>
    </>
  );
}

export default AvatarDemo;
