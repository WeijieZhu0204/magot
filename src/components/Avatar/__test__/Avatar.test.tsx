import React from 'react';
import { render } from 'enzyme';
import Avatar from '../Avatar';
import { injectStyleTestSuites } from '../../../../test/shared.test';

describe('Avatar', () => {
  it('should renders dom correctly', () => {
    const wrapper = render(<Avatar />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should renders a image Avatar', () => {
    const wrapper = render(<Avatar src="../avatar.png" />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.hasClass('mgt-avatar-image')).toBe(true);
    expect(wrapper.find('img')).toHaveLength(1);
  });

  it('should renders a text Avatar', () => {
    const wrapper = render(<Avatar text="USER" />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.hasClass('mgt-avatar-text')).toBe(true);
    expect(wrapper.find('.text')).toHaveLength(1);
  });

  it('should renders a icon Avatar', () => {
    const wrapper = render(<Avatar icon="user" />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.hasClass('mgt-avatar-icon')).toBe(true);
    expect(wrapper.find('.icon')).toHaveLength(1);
  });

  it('should renders `radius` prop when passed in', () => {
    const wrapper1 = render(<Avatar radius={4} />);
    const wrapper2 = render(<Avatar radius="10px" />);
    expect(wrapper1).toMatchSnapshot();
    expect(wrapper2).toMatchSnapshot();
    expect(wrapper1.attr('style')).toEqual('border-radius:4px');
    expect(wrapper2.attr('style')).toEqual('border-radius:10px');
  });

  it('should renders `srcset` prop when passed in', () => {
    const wrapper = render(
      <Avatar src="../avatar.png" srcset="../avatar2.png" />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('img').attr('srcset')).toEqual('../avatar2.png');
  });

  it('should renders `alt` prop when passed in', () => {
    const wrapper = render(<Avatar src="../avatar.png" alt="AVATAR" />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('img').attr('alt')).toEqual('AVATAR');
  });

  it('should renders `size` prop when passed in', () => {
    const wrapper1 = render(<Avatar size="small" />);
    const wrapper2 = render(<Avatar size="large" />);
    expect(wrapper1).toMatchSnapshot();
    expect(wrapper2).toMatchSnapshot();
    expect(wrapper1.hasClass('mgt-avatar-sm')).toBe(true);
    expect(wrapper2.hasClass('mgt-avatar-lg')).toBe(true);
  });

  injectStyleTestSuites(Avatar);
});
