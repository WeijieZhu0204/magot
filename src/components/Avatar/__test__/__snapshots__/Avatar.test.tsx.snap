// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should renders \`alt\` prop when passed in 1`] = `
<div
  class="mgt-avatar mgt-avatar-image"
  style="border-radius:50%"
>
  <img
    alt="AVATAR"
    class="image"
    src="../avatar.png"
  />
</div>
`;

exports[`Avatar should renders \`radius\` prop when passed in 1`] = `
<div
  class="mgt-avatar mgt-avatar-icon"
  style="border-radius:4px"
>
  <i
    class="mgt-icon mgt-icon-user icon"
  />
</div>
`;

exports[`Avatar should renders \`radius\` prop when passed in 2`] = `
<div
  class="mgt-avatar mgt-avatar-icon"
  style="border-radius:10px"
>
  <i
    class="mgt-icon mgt-icon-user icon"
  />
</div>
`;

exports[`Avatar should renders \`size\` prop when passed in 1`] = `
<div
  class="mgt-avatar mgt-avatar-sm mgt-avatar-icon"
  style="border-radius:50%"
>
  <i
    class="mgt-icon mgt-icon-user icon"
  />
</div>
`;

exports[`Avatar should renders \`size\` prop when passed in 2`] = `
<div
  class="mgt-avatar mgt-avatar-lg mgt-avatar-icon"
  style="border-radius:50%"
>
  <i
    class="mgt-icon mgt-icon-user icon"
  />
</div>
`;

exports[`Avatar should renders \`srcset\` prop when passed in 1`] = `
<div
  class="mgt-avatar mgt-avatar-image"
  style="border-radius:50%"
>
  <img
    class="image"
    src="../avatar.png"
    srcset="../avatar2.png"
  />
</div>
`;

exports[`Avatar should renders a icon Avatar 1`] = `
<div
  class="mgt-avatar mgt-avatar-icon"
  style="border-radius:50%"
>
  <i
    class="mgt-icon mgt-icon-user icon"
  />
</div>
`;

exports[`Avatar should renders a image Avatar 1`] = `
<div
  class="mgt-avatar mgt-avatar-image"
  style="border-radius:50%"
>
  <img
    class="image"
    src="../avatar.png"
  />
</div>
`;

exports[`Avatar should renders a text Avatar 1`] = `
<div
  class="mgt-avatar mgt-avatar-text"
  style="border-radius:50%"
>
  <span
    class="text"
    style="transform:scale(1) translateX(-50%)"
  >
    USER
  </span>
</div>
`;

exports[`Avatar should renders custom className 1`] = `
<div
  class="mgt-avatar mgt-avatar-icon jest-test-class"
  style="border-radius:50%"
>
  <i
    class="mgt-icon mgt-icon-user icon"
  />
</div>
`;

exports[`Avatar should renders custom styles 1`] = `
<div
  class="mgt-avatar mgt-avatar-icon"
  style="border-radius:50%;color:#333;background-color:#f0f0f0"
>
  <i
    class="mgt-icon mgt-icon-user icon"
  />
</div>
`;

exports[`Avatar should renders dom correctly 1`] = `
<div
  class="mgt-avatar mgt-avatar-icon"
  style="border-radius:50%"
>
  <i
    class="mgt-icon mgt-icon-user icon"
  />
</div>
`;
